matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -161,84 +166,201 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
-        queryGraph.lookup(allCandidates);
+        // create candidates from the entries in the graph (a candidate is basically an entry + direction):
+        List<TimeStep<GPXExtension, GPXEntry, Path>> timeSteps =
+                createTimeSteps(filteredGPXEntries, queriesPerEntry, queryGraph);
+        logger.info("=============== Time steps ===============");
+        i = 1;
+        for (TimeStep<GPXExtension, GPXEntry, Path> ts : timeSteps) {
+            logger.info("Candidates for time step {}", i++);
+            for (GPXExtension candidate : ts.candidates) {
+                logger.info(candidate.toString());
+            }
+        }
 
+        // Compute the most likely sequence of map matching candidates:
         List<SequenceState<GPXExtension, GPXEntry, Path>> seq = computeViterbiSequence(timeSteps,
-                gpxList, queryGraph);
+                gpxList.size(), queryGraph);
+
+        logger.info("=============== Viterbi results =============== ");
+        i = 1;
+        for (SequenceState<GPXExtension, GPXEntry, Path> ss : seq) {
+            logger.info("{}: {}, path: {}", i, ss.state,
+                    ss.transitionDescriptor != null ? ss.transitionDescriptor.calcEdges() : null);
+            i++;
+        }
 
+        // finally, extract the result:
         final EdgeExplorer explorer = queryGraph.createEdgeExplorer(edgeFilter);
-        MatchResult matchResult = computeMatchResult(seq, gpxList, allCandidates, explorer);
+
+        // Needs original gpxList to compute stats.
+        MatchResult matchResult = computeMatchResult(seq, gpxList, queriesPerEntry, explorer);
+
+        logger.info("=============== Matched real edges =============== ");
+        i = 1;
+        for (EdgeMatch em : matchResult.getEdgeMatches()) {
+            logger.info("{}: {}", i, em.getEdgeState());
+            i++;
+        }
 
         return matchResult;
     }
+    
+    /**
+     * Filters GPX entries to only those which will be used for map matching (i.e. those which
+     * are separated by at least 2 * measurementErrorSigman
+     */
+    private List<GPXEntry> filterGPXEntries(List<GPXEntry> gpxList) {
+        List<GPXEntry> filtered = new ArrayList<>();
+        GPXEntry prevEntry = null;
+        int last = gpxList.size() - 1;
+        for (int i = 0; i <= last; i++) {
+            GPXEntry gpxEntry = gpxList.get(i);
+            if (i == 0 || i == last || distanceCalc.calcDist(
+                    prevEntry.getLat(), prevEntry.getLon(),
+                    gpxEntry.getLat(), gpxEntry.getLon()) > 2 * measurementErrorSigma) {
+                filtered.add(gpxEntry);
+                prevEntry = gpxEntry;
+            } else {
+                logger.info("Filter out GPX entry: {}", i + 1);
+            }
+        }
+        return filtered;
+    }
+
+    /**
+     * Find the possible locations of each qpxEntry in the graph.
+     */
+    private List<List<QueryResult>> findGPXEntriesInGraph(List<GPXEntry> gpxList,
```

Comment:
It should be named like findQueryResults or lookupGPXEntries, findGPXEntries implies that we return something like `List<GPXEntry>` (?)